R/florabank_queries.R

```diff
@@ -0,0 +1,424 @@
+#' scient_name = 'Dactylorhiza maculata (L.) Soó')
+#'
+#' Dactmacu2 <-	florabank_observations(db_connectie,
+#' dutch_name = 'Gevlekte orchis')
+#'
+#' all.equal(Dactmacu1, Dactmacu2)
+#'
+#' # use wildcards to retrieve all partial matches to a name
+#' Dactmacu3 <-	florabank_observations(db_connectie,
+#' scient_name = 'Dactylorhiza maculata%')
+#'
+#' # disconnect from florabank
+#' dbDisconnect(db_connectie)
+#' }
+
+florabank_observations <- function(connection, scient_name, dutch_name) {
+
+  assert_that(class(connection)[1] == "Microsoft SQL Server",
+              msg = "Not a connection object to database.")
+  assert_that(connection@info$dbname == "D0021_00_userFlora")
+
+  if (missing(scient_name) & missing(dutch_name)) {
+    stop("Please provide either a scientific name or a Dutch species name.")
+  }
+
+  if (!missing(scient_name) & !missing(dutch_name)) {
+    stop("Both a scientific name and a Dutch name are provided. Use either,
+         but not both.")
+  }
+
+  if (!missing(scient_name)) {
+    glue_statement <- glue_sql(
+      "SELECT DISTINCT
+      tblTaxon.NaamNederlands
+      , tblTaxon.NaamWetenschappelijk
+      , cdeBron.Omschrijving AS Bron
+      , tblWaarneming.BeginDatum
+      , tblWaarneming.EindDatum
+      , tblIFBLHok.Code AS hok
+      , tblWaarneming.Opmerking AS Toponiem
+      , tblMeting.CommentaarTaxon
+      , tblMeting.CommentaarHabitat
+      , tblMedewerker.Voornaam
+      , tblMedewerker.Achternaam
+      , tblWaarneming.ID
+      , tblWaarneming.Cor_X AS X_waarneming
+      , tblWaarneming.Cor_Y AS Y_waarneming
+      , tblMeting.Cor_X AS X_meting
+      , tblMeting.Cor_Y AS Y_meting
+      FROM dbo.tblWaarneming
+      INNER JOIN dbo.tblMeting ON tblWaarneming.ID = tblMeting.WaarnemingID
+      INNER JOIN dbo.relTaxonTaxon ON relTaxonTaxon.TaxonIDChild = tblMeting.TaxonID
+      INNER JOIN dbo.tblTaxon ON tblTaxon.ID = relTaxonTaxon.TaxonIDParent
+      INNER JOIN dbo.relTaxonTaxonGroep ON relTaxonTaxonGroep.TaxonID = tblTaxon.ID
+      INNER JOIN dbo.tblIFBLHok ON tblIFBLHok.ID = tblWaarneming.IFBLHokID
+      INNER JOIN dbo.relWaarnemingMedewerker ON relWaarnemingMedewerker.WaarnemingID = tblWaarneming.ID
+      INNER JOIN dbo.tblMedewerker ON tblMedewerker.ID = relWaarnemingMedewerker.MedewerkerID
+      INNER JOIN dbo.cdeBron ON cdeBron.Code = tblWaarneming.BronCode
+      WHERE 1=1
+      AND (tblMeting.MetingStatusCode='GDGA' OR tblMeting.MetingStatusCode='GDGK')
+      AND tblTaxon.NaamWetenschappelijk LIKE {scient_name}
+      ORDER BY tblWaarneming.BeginDatum DESC;",
+      scient_name = scient_name,
+      .con = connection)
+    glue_statement <- iconv(glue_statement, from =  "UTF-8", to = "latin1")
+    query_result <- dbGetQuery(connection, glue_statement)
+    query_result <- as_tibble(query_result)
+    return(query_result)
+  }
+
+  if (!missing(dutch_name)) {
+    glue_statement <- glue_sql(
+      "SELECT DISTINCT
+      tblTaxon.NaamNederlands
```

Comment:
nearly identical SQL as above, only the WHERE is slightly different. The code is easier to maintain if you combine both statements. E.g. put the common part in an object.

Reply:
Resolved in 172113e